# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/ipc/GamepadMessageUtils.h
# Commit: ec6a64b639a8
# Full Hash: ec6a64b639a82e957376c09a71aef7ce62b3a5aa
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-12-03 09:47:26
# Regressor Bug: 1657404
# File Overlap Count: 3
# Description:
#   Bug 1657404 - Implement a strongly-typed, service-dependent gamepad handle r=handyman,aklotz
#   
#   Currently, the gamepad code uses a uint32_t as a handle and does some trickery
#   with it by trying to create a unique ID and adding an offset to it for VR code.
#   
# ==============================================================================

diff -r 6792f28b99bd -r ec6a64b639a8 dom/gamepad/ipc/GamepadMessageUtils.h
--- a/dom/gamepad/ipc/GamepadMessageUtils.h	Wed Dec 02 22:36:35 2020 +0000
+++ b/dom/gamepad/ipc/GamepadMessageUtils.h	Wed Dec 02 23:06:05 2020 +0000
@@ -9,9 +9,9 @@
 
 #include "ipc/IPCMessageUtils.h"
 #include "mozilla/dom/GamepadBinding.h"
+#include "mozilla/dom/GamepadHandle.h"
 #include "mozilla/dom/GamepadLightIndicatorBinding.h"
 #include "mozilla/dom/GamepadPoseState.h"
-#include "mozilla/dom/GamepadServiceType.h"
 #include "mozilla/dom/GamepadTouchState.h"
 
 namespace IPC {
@@ -38,13 +38,6 @@
           mozilla::dom::GamepadHand(mozilla::dom::GamepadHand::EndGuard_)> {};
 
 template <>
-struct ParamTraits<mozilla::dom::GamepadServiceType>
-    : public ContiguousEnumSerializer<
-          mozilla::dom::GamepadServiceType, mozilla::dom::GamepadServiceType(0),
-          mozilla::dom::GamepadServiceType(
-              mozilla::dom::GamepadServiceType::NumGamepadServiceType)> {};
-
-template <>
 struct ParamTraits<mozilla::dom::GamepadCapabilityFlags>
     : public BitFlagsEnumSerializer<
           mozilla::dom::GamepadCapabilityFlags,
@@ -138,6 +131,27 @@
   }
 };
 
+template <>
+struct ParamTraits<mozilla::dom::GamepadHandleKind>
+    : public ContiguousEnumSerializerInclusive<
+          mozilla::dom::GamepadHandleKind,
+          mozilla::dom::GamepadHandleKind::GamepadPlatformManager,
+          mozilla::dom::GamepadHandleKind::VR> {};
+
+template <>
+struct ParamTraits<mozilla::dom::GamepadHandle> {
+  typedef mozilla::dom::GamepadHandle paramType;
+  static void Write(Message* aMsg, const paramType& aParam) {
+    WriteParam(aMsg, aParam.mValue);
+    WriteParam(aMsg, aParam.mKind);
+  }
+  static bool Read(const Message* aMsg, PickleIterator* aIter,
+                   paramType* aParam) {
+    return ReadParam(aMsg, aIter, &aParam->mValue) &&
+           ReadParam(aMsg, aIter, &aParam->mKind);
+  }
+};
+
 }  // namespace IPC
 
 #endif  // mozilla_dom_gamepad_GamepadMessageUtils_h